Entrance tracker tweaks

Malkierian opened the pull request:
Hides return entrance transitions for grotto, interior, and dungeon types when "Decouple entrances" is off.

Also highlights the inverse of the last entrance when checking for "Highlight last entrance" if the actual last entrance is hidden on the tracker due to the first part of this pull request, if Decouple is off, and checks for this first to allow for highlighting "last entrance" in an area Link is currently in.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/561206585.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/561206586.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/561206587.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/561206588.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/561206589.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/561206591.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Hide return transitions for dungeon, grotto and interior entrances when "Decouple entrances" is off.

Malkierian committed:
"Highlight last entrance" now highlights the override reverseIndex entry in the list if "Decouple entrances" is off, to correspond with not displaying return directions for the same setting (without this, nothing gets highlighted with the return transitions hidden with "Decouple entrances" off).

Malkierian committed:
Removed unnecessary printf

Archez commented on soh/soh/Enhancements/randomizer/randomizer_entrance_tracker.cpp:
```diff
@@ -831,6 +831,10 @@ void DrawEntranceTracker(bool& open) {
 
+                    if (original->type == ENTRANCE_TYPE_DUNGEON || original->type == ENTRANCE_TYPE_GROTTO || original->type == ENTRANCE_TYPE_INTERIOR)
+                        if (original->oneExit != 1 && OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_DECOUPLED_ENTRANCES) == RO_GENERIC_OFF)
+                            continue;
```
We generally do not like bracket-less if statements, as they can introduce errors pretty easily. This can be done with one if statement overall (the expression can be on multiple lines).

And can you add a comment clarifying what this is doing.

Also this should be moved in the loop section above as thats used for filtering and will populate the underlying `displayEntrances` array used in this loop (this will also make sure the "undiscovered" count when using "Hide undiscovered" is reported correctly).

Malkierian replied:
Ah, yes, that loop would make more sense. If I were to make a setting for it, what should I put in as the text? I was thinking "Show Redundant Entrances", but while it's descriptive, it seems unnecessarily long.

Archez replied:
Maybe "Hide reverse", with a tooltip that says "Hide reversed entries for two-way entrances when decoupled entrances is turned off".

We can make the default for the CVar be on so that its defaulted to true, and then the checkbox can be disabled if decoupled is on. There is a few examples of how to mark a checkbox as disabled in the randomizer.cpp

Malkierian replied:
I got the disabling just fine, but since the Imgui element used is the `EnhancementCheckbox`, how do I set the default to checked? I feel like I'm missing code for saving/loading CVars.

Malkierian replied:
Nevermind, looks like it just needed a toggle to remember its state across loads.

Malkierian replied:
Think it's all done.

Archez commented on soh/soh/Enhancements/randomizer/randomizer_entrance_tracker.cpp:
```diff
@@ -846,7 +850,11 @@ void DrawEntranceTracker(bool& open) {
                     // Handle highlighting and auto scroll
-                    if (LinkIsInArea(original) != -1) {
+                    if (original->index == lastEntranceIndex ||
+                        (override->reverseIndex == lastEntranceIndex && OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_DECOUPLED_ENTRANCES) == RO_GENERIC_OFF) &&
+                            CVarGetInteger("gEntranceTrackerHighlightPrevious", 0)) {
```
Just a nit-pick, but could you wrap the two OR-ed statements here in parentheses

Malkierian replied:
I don't know why this comment had three different lines of code represented between the two changes, but I implemented it in my update.

Archez replied:
I messed up the comment/suggestion, and didn't want to redo it so I just removed the "suggestion". Thank you 🙂

Archez reviewed (commented):
Overall looks good. Left a few suggestions.

One thing I am torn on is if this behavior change should be controlled via a toggle in the tracker settings (could be defaulted to on). Not sure if others may prefer the original behavior.

Malkierian committed:
Moved redundancy check to filtering loop instead of display loop.

Introduced setting to optionally force showing redundant entrances when decoupled is off.

Formatting changes suggested by Archez.

Malkierian committed:
Changed "show redundant" to "hide reverse" where applicable, and change

Added option disabling based on Decouple Entrances.

Malkierian committed:
Finished descriptive comment.

Archez commented on soh/soh/Enhancements/randomizer/randomizer_entrance_tracker.cpp:
```diff
@@ -687,6 +687,12 @@ void DrawEntranceTracker(bool& open) {
+                static const char* disableHideReverseEntrancesText = "This option is disabled because \"Decouple Entrances\" is enabled.";
+                UIWidgets::EnhancementCheckbox("Hide reverse", "gEntranceTrackerHideReverseEntrances",
+                                              disableHideReverseEntrances, disableHideReverseEntrancesText);                
+                UIWidgets::Tooltip("Hide reverse entrance transitions when Decouple Entrances is off");
```
If you use `UIWidgets::PaddedEnhancementCheckbox` you don't need the spacer, and it excepts the same params, plus extra params for adding a top or bottom space.

The last param allows you to specify the default value. Check the method signature to see what I mean.

I think this checkbox also should go on the left column since that contains the settings related to the list.

Malkierian replied:
Yeah, that makes sense. Done.

Malkierian committed:
Moved "Hide reverse" to left column of tracker settings.

Changed to PaddedEnhancementCheckbox, changed checkbox padding accordingly, and set default to true.

Archez commented on soh/soh/Enhancements/randomizer/randomizer_entrance_tracker.cpp:
```diff
@@ -669,8 +669,13 @@ void DrawEntranceTracker(bool& open) {
                 UIWidgets::Tooltip("Collapse undiscovered entrances towards the bottom of each group");
+                bool disableHideReverseEntrances = OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_DECOUPLED_ENTRANCES) == RO_GENERIC_ON;
+                static const char* disableHideReverseEntrancesText = "This option is disabled because \"Decouple Entrances\" is enabled.";
+                UIWidgets::PaddedEnhancementCheckbox("Hide reverse", "gEntranceTrackerHideReverseEntrances", true, true,
+                                              disableHideReverseEntrances, disableHideReverseEntrancesText, UIWidgets::CheckboxGraphics::Cross, false);                
```
One last suggestion: (I just found this out), but if the `PaddedEnhancementCheckbox` has a bottom padding, then the tooltip doesn't render. We can just set it to false for now to get the tooltip to render (not a big deal)

Also we want the default value to be true (matching the default value passed into CVarGet below)

```suggestion
                UIWidgets::PaddedEnhancementCheckbox("Hide reverse", "gEntranceTrackerHideReverseEntrances", true, false,
                                              disableHideReverseEntrances, disableHideReverseEntrancesText, UIWidgets::CheckboxGraphics::Cross, true);
```

Malkierian replied:
LMAO, how does that even work? Are you OK with the separator line underneath being that much closer, or should we have a spacer in there?

And how did I miss that default setting? *sigh*

Archez replied:
Can always manually add the `UIWidgets:Spacer(0);` underneath the tooltip for now to add back the spacing.

Basically the tooltip applies to the last ImGui element, and if bottom spacing is added to the PaddedEnhancementCheckbox, then the tooltip is applied to the space, rather than the checkbox.

Probably something we can look into in another PR, but for now a manual space here is fine.

Malkierian replied:
Done.

Malkierian committed:
Restore tooltip to Hide Reverse option, fix default value.

Co-authored-by: Adam Bird <Archez@users.noreply.github.com>

Malkierian committed:
Restored padding underneath last option in entrance tracker List Items section.

Archez reviewed (approved):
Alright LGTM 🙂 thanks for working through this with me!

Malkierian commented:
Glad to get more insight into the processes, hopefully the next PRs will be a little cleaner now.
